src/runtime/hexagon/hexagon_buffer_manager.h

```diff
@@ -82,28 +87,35 @@ class HexagonBufferManager {
 
   //! \brief Returns whether the HexagonBufferManager has any allocations.
   bool empty() {
-    std::lock_guard<std::mutex> lock(map_mutex_);
+    std::lock_guard<std::mutex> lock(mutex_);
     return hexagon_buffer_map_.empty();
   }
 
-  //! \brief Returns a vector of currently allocated pointers, owned by the manager.
-  // Note - this should only be used by the device API to keep track of what
-  // was in the manager when HexagonDeviceAPI::ReleaseResources is called.
-  std::vector<void*> current_allocations() {
-    std::vector<void*> allocated;
-    std::lock_guard<std::mutex> lock(map_mutex_);
+  void Reset() {
+    std::lock_guard<std::mutex> lock(mutex_);
+    hexagon_buffer_map_.clear();
+    released_buffers_.clear();
+  }
+
+  void Release() {
     for (const auto& [data_ptr, buffer] : hexagon_buffer_map_) {
-      allocated.push_back(data_ptr);
+      FreeHexagonBuffer(data_ptr);
+      {
+        std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
Looking at it again, because `FreeHexagonBuffer` may modify `hexagon_buffer_map_`, the loop iterator would be invalidated even if we aren't pre-empted by another thread.  When `unordered_map::erase` is called, [only the iterator](https://en.cppreference.com/w/cpp/container/unordered_map#Iterator_invalidation) to the erased element is invalidated.  However, that is exactly the iterator we're holding, and what we need to increment for the loop.  The compiler isn't required to diagnose invalidated iterators, so we're into undefined behavior at that point.

I think the following implementation would work better, both to avoid lock release/re-acquire, and would avoid modification of `hexagon_buffer_map_` while iterating over it.

```c++
void Release() {
  std::lock_guard<std::mutex> lock(mutex_);
  for(const auto& [data_ptr, buffer] : hexagon_buffer_map_) {
    released_buffers_.insert(data_ptr);
  }
  hexagon_buffer_map_.clear();
}
```

Reply:
Good call.  Thanks for this review.